File: src/main/java/gargoyle/netsquares/store/prop/BasePrefProperty.java
package gargoyle.netsquares.store.prop;

import gargoyle.netsquares.prop.base.BaseRWProperty;

import java.util.Objects;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

public abstract class BasePrefProperty<T> extends BaseRWProperty<T> {
    protected final String key;
    protected final Preferences preferences;
    private final transient T defaultValue;

    protected BasePrefProperty(Class<?> pkg, String key, T defaultValue) {
        this.key = Objects.requireNonNull(key);
        preferences = Preferences.userNodeForPackage(pkg);
        this.defaultValue = defaultValue;
    }

    protected final void flush() {
        try {
            preferences.sync();
        } catch (BackingStoreException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    public abstract T get(T defaultValue);

    public final T get() {
        return get(defaultValue);
    }

    public T getValue() {
        T t = get(defaultValue);
        return t == null ? nonNull() : t;
    }

    @SuppressWarnings("unchecked")
    private T nonNull() {
        try {
            return ((Class<T>) defaultValue.getClass()).getConstructor().newInstance();
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }
}

File: src/main/kotlin/gargoyle/netsquares/store/prop/BasePrefProperty.kt
package gargoyle.netsquares.store.prop

import gargoyle.netsquares.prop.base.BaseRWProperty
import java.util.prefs.Preferences
import kotlin.reflect.KClass

abstract class BasePrefProperty<T : Any> protected constructor(pkg: KClass<*>, key: String, defaultValue: T) :
    BaseRWProperty<T>() {
    protected val key: String
    protected val preferences: Preferences

    private val defaultValue: T

    init {
        this.key = key
        preferences = Preferences.userNodeForPackage(pkg.java)
        this.defaultValue = defaultValue
    }

    override fun flush() {
        try {
            preferences.sync()
        } catch (e: Exception) {
            throw RuntimeException(e.message, e)
        }
    }

    override fun get(): T {
        return get(defaultValue)
    }

    abstract operator fun get(defaultValue: T): T
    val value: T
        get() {
            return get(defaultValue)
        }
}
